newlib/libc/machine/riscv/memcpy-asm.S

```diff
@@ -9,24 +9,110 @@
    http://www.opensource.org/licenses.
 */
 
-#if defined(PREFER_SIZE_OVER_SPEED) || defined(__OPTIMIZE_SIZE__)
 .text
 .global memcpy
 .type	memcpy, @function
 memcpy:
-  mv t1, a0
+#if defined(PREFER_SIZE_OVER_SPEED) || defined(__OPTIMIZE_SIZE__)
+  mv a3, a0
   beqz a2, 2f
 
 1:
-  lb t2, 0(a1)
-  sb t2, 0(t1)
+  lbu a4, 0(a1)
+  sb a4, 0(a3)
   add   a2, a2, -1
-  add   t1, t1, 1
+  add   a3, a3, 1
   add   a1, a1, 1
   bnez a2, 1b
 
 2:
   ret
 
-  .size	memcpy, .-memcpy
+#else
+.Lwordify:
+    andi   a3, a0, 3
+    beqz   a3, .Laligned
+    li    a3, 3
+    bleu  a2, a3, .Lcopy1
+
+    lbu    a3, 0(a1)
+    sb     a3, 0(a0)
+    addi   a0, a0, 1
+    addi   a1, a1, 1
+    addi   a2, a2, -1
+    j      .Lwordify
+
+.Laligned:
+#ifdef __riscv_misaligned_avoid
+    andi   a3, a1, 3
+    bnez   a3, .Lcopy1
 #endif
+
+.Lcopy32:
+    li     a3, 32
+    blt    a2, a3, .Lcopy16
+    lw     t1, 0(a1)
+    lw     t2, 4(a1)
+    lw     t3, 8(a1)
+    lw     a3, 12(a1)
+    lw     a4, 16(a1)
+    lw     a5, 20(a1)
+    lw     a6, 24(a1)
+    lw     a7, 28(a1)
+
+    sw     t1, 0(a0)
+    sw     t2, 4(a0)
+    sw     t3, 8(a0)
+    sw     a3, 12(a0)
+    sw     a4, 16(a0)
+    sw     a5, 20(a0)
+    sw     a6, 24(a0)
+    sw     a7, 28(a0)
+
+    addi   a1, a1, 32
+    addi   a0, a0, 32
+    addi   a2, a2, -32
+    j      .Lcopy32
+
+.Lcopy16:
+    li     a3, 16
+    blt    a2, a3, .Lcopy4
+    lw     a3, 0(a1)
+    lw     a4, 4(a1)
+    lw     a5, 8(a1)
+    lw     a6, 12(a1)
+
+    sw     a3, 0(a0)
+    sw     a4, 4(a0)
+    sw     a5, 8(a0)
+    sw     a6, 12(a0)
+
+    addi   a1, a1, 16
+    addi   a0, a0, 16
+    addi   a2, a2, -16
+    j      .Lcopy16
+
+.Lcopy4:
+    li     a3, 4
+    blt    a2, a3, .Lcopy1
```

Comment:
why are we looping here? if you are just copying the remainder, there should just be one word sized datum be left to copy. This also leads to a simplification where you do not really need to keep decrementing the remaining bytes, because it becomes a fixed sequence.

Reply:
I am jumping from copy32 to copy4. so we can get into copy4 with 31 bytes remaining. Having copy32, copy16, copy8, and copy4 doesn't give any extra perf